pkg/operator/configobserver/config_observer_controller.go

```diff
@@ -105,6 +106,17 @@ func (c ConfigObserver) sync() error {
 		}
 	}
 
+	reverseMergedObservedConfig := map[string]interface{}{}
+	for i := len(observedConfigs) - 1; i >= 0; i-- {
+		if err := mergo.Merge(&reverseMergedObservedConfig, observedConfigs[i]); err != nil {
+			glog.Warningf("merging observed config failed: %v", err)
+		}
+	}
+
+	if !equality.Semantic.DeepEqual(mergedObservedConfig, reverseMergedObservedConfig) {
+		errs = append(errs, errors.New("non-deterministic config observation detected"))
```

Comment:
what is a non-deterministic config observation, what can a observer developer do, what can an admin do?

Reply:
If I understand the code correctly, this happens if two observers override eachothers' values. Only one ends up in the config, depending on order. Can we put that into some understandable words?

I don't particularly like the whole mechanism implemented here. If there are conflicts, they might depend on user input. Because there is zero hint for the reason, this is a ticking time bomb. They operator becomes blocked, the cluster unusable in the worst case.

/cc @deads2k

Reply:
An idea: in linear time we should be able to find the conflicting observers. Can we print them and their respective patch for the config?

Reply:
I agree something smells here, I had to drop `mergo.WithOverride` in the previous `mergo.Merge` call which seems suspicious. @sanchezl this definitely needs more thought...

Reply:
I've meant in https://github.com/openshift/library-go/pull/78

Reply:
> what is a non-deterministic config observation, what can a observer developer do, what can an admin do?

It means that two different configobservation methods are setting the same field to different values.  This means that there is a developer error and a bug that needs to be fixed.  This should prevent them from merging.  We don't have any we know of yet.

Reply:
> An idea: in linear time we should be able to find the conflicting observers. Can we print them and their respective patch for the config?

Not statically, but at runtime you could.  Given that it's a coder error that shouldn't be able to merge anymore (this makes status failing), I don't think it's urgent.

Reply:
> I don't particularly like the whole mechanism implemented here. If there are conflicts, they might depend on user input. Because there is zero hint for the reason, this is a ticking time bomb. The operator becomes blocked, the cluster unusable in the worst case.

They might, but I think that straight failing is better than having a flapping configuration as we call this in a control loop.